src/GitHub.App/Services/TeamExplorerContext.cs

```diff
@@ -0,0 +1,221 @@
+                {
+                    return repo.Head.TrackedBranch?.Tip.Sha;
+                }
+            }
+
+            public ILocalRepositoryModel CreateRepository(string path) => new LocalRepositoryModel(path);
+        }
+
+        void Refresh(object gitExt)
+        {
+            try
+            {
+                var repo = FindActiveRepository(gitExt);
+                var newSolutionPath = dte?.Solution?.FullName;
+
+                if (repo == null && newSolutionPath == solutionPath)
+                {
+                    // Ignore when ActiveRepositories is empty and solution hasn't changed.
+                    // https://github.com/github/VisualStudio/issues/1421
+                    log.Information("Ignoring no ActiveRepository when solution hasn't changed");
+                }
+                else
+                {
+                    var newRepositoryPath = repo?.RepositoryPath;
+                    var newBranchName = repo?.BranchName;
+                    var newHeadSha = repo?.HeadSha;
+                    var newTrackedSha = newRepositoryPath != null ? service.FindTrackedSha(newRepositoryPath) : null;
+
+                    if (newRepositoryPath != repositoryPath)
+                    {
+                        log.Information("Fire PropertyChanged event for ActiveRepository");
+                        ActiveRepository = repo != null ? service.CreateRepository(repo.RepositoryPath) : null;
+                        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(ActiveRepository)));
+                    }
+                    else if (newBranchName != branchName)
+                    {
+                        log.Information("Fire StatusChanged event when BranchName changes for ActiveRepository");
+                        StatusChanged?.Invoke(this, EventArgs.Empty);
+                    }
+                    else if (newHeadSha != headSha)
+                    {
+                        log.Information("Fire StatusChanged event when HeadSha changes for ActiveRepository");
+                        StatusChanged?.Invoke(this, EventArgs.Empty);
+                    }
+                    else if (newTrackedSha != trackedSha)
+                    {
+                        log.Information("Fire StatusChanged event when TrackedSha changes for ActiveRepository");
+                        StatusChanged?.Invoke(this, EventArgs.Empty);
+                    }
+
+                    repositoryPath = newRepositoryPath;
+                    branchName = newBranchName;
+                    headSha = newHeadSha;
+                    solutionPath = newSolutionPath;
+                    trackedSha = newTrackedSha;
+                }
+            }
+            catch (Exception e)
+            {
+                log.Error(e, "Refreshing active repository");
+            }
+        }
+
+        static RepositoryInfo FindActiveRepository(object gitExt)
```

Comment:
How would this be unit tested?

Reply:
It does have lots of tests. I pass in an object with a similar "shape". See `FakeGitExt`.